### PR TITLE
ERC-404 basic support

### PR DESCRIPTION
**Pull request by vbaranov**
Resolves https://github.com/blockscout/blockscout/issues/9378

## Motivation

Basic support of [ERC-404](https://github.com/Pandora-Labs-Org/erc404).

## Changelog

Define ERC-404 tokens as a separate token types. Detect balance owners of related NFTs as well as track balances as of ERC-20 token.

## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [x] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [x] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [x] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex [hunk 1]
```diff
@@ -40,7 +40,8 @@ defmodule BlockScoutWeb.Tokens.Helper do
   end
 
   # TODO: remove this clause along with token transfer denormalization
-  defp do_token_transfer_amount(%Token{type: "ERC-20", decimals: nil}, nil, amount, _amounts, _token_ids) do
+  defp do_token_transfer_amount(%Token{type: type, decimals: nil}, type, amount, _amounts, _token_ids)
```
Comment (Qwerty5Uiop) [inline apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:43]: This clause is needed for cases when token transfers denormalization is not finished yet, so this line should look like this (second parameter is `nil`). And the next clause should match on `ERC-404` as well
```suggestion
  defp do_token_transfer_amount(%Token{type: type, decimals: nil}, nil, amount, _amounts, _token_ids)
```

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex [hunk 2]
```diff
@@ -49,7 +50,8 @@ defmodule BlockScoutWeb.Tokens.Helper do
   end
 
   # TODO: remove this clause along with token transfer denormalization
-  defp do_token_transfer_amount(%Token{type: "ERC-20", decimals: decimals}, nil, amount, _amounts, _token_ids) do
+  defp do_token_transfer_amount(%Token{type: type, decimals: decimals}, nil, amount, _amounts, _token_ids)
+       when type in ["ERC-20", "ERC-404"] do
     {:ok, CurrencyHelper.format_according_to_decimals(amount, decimals)}
   end
```
Comment (Qwerty5Uiop) [inline apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:57]: Same for the next line

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex [hunk 3]
```diff
@@ -40,7 +40,8 @@ defmodule BlockScoutWeb.Tokens.Helper do
...
```
Comment (Qwerty5Uiop) [inline apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:40]: If we consider `ERC-404` transfers amount as `ERC-20`, we should add this type to functions above as well

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex
Comment (Qwerty5Uiop) [inline apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:1]: Also, seems like functions `do_token_transfer_amount_for_api/5` should match to `ERC-404` as well

File: apps/explorer/lib/explorer/chain/token/instance.ex [hunk 4]
```diff
@@ -113,6 +113,10 @@ defmodule Explorer.Chain.Token.Instance do
     erc_1155_token_instances_by_address_hash(address_hash, options)
   end
 
+  defp nft_list(address_hash, ["ERC-404"], options) do
+    erc_404_token_instances_by_address_hash(address_hash, options)
+  end
+
   defp nft_list(address_hash, _, options) do
     paging_options = Keyword.get(options, :paging_options, Chain.default_paging_options())
```
Comment (Qwerty5Uiop) [inline apps/explorer/lib/explorer/chain/token/instance.ex:122]: Shouldn't be there case for `%PagingOptions{key: {_contract_address_hash, _token_id, "ERC-404"}}` along with `%PagingOptions{key: {_contract_address_hash, _token_id, "ERC-1155"}}`?

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex (hunk 2 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:57]: Processed in https://github.com/blockscout/blockscout/pull/9407/commits/07b9fc41820585c9e4e550323a57725e56e944bf

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex (hunk 3 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:40]: Processed in https://github.com/blockscout/blockscout/pull/9407/commits/07b9fc41820585c9e4e550323a57725e56e944bf

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:1]: Processed in https://github.com/blockscout/blockscout/pull/9407/commits/07b9fc41820585c9e4e550323a57725e56e944bf

File: apps/explorer/lib/explorer/chain/token/instance.ex (hunk 4 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/chain/token/instance.ex:122]: Processed in https://github.com/blockscout/blockscout/pull/9407/commits/07b9fc41820585c9e4e550323a57725e56e944bf

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex (hunk 2 above)
Comment (Qwerty5Uiop) [reply apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:57]: There is still missing update on line 58

File: apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex [hunk 5]
```diff
@@ -31,16 +31,17 @@ defmodule BlockScoutWeb.Tokens.Helper do
...
+  defp do_token_transfer_amount(_token, type, nil, nil, _token_ids) when type in ["ERC-20", "ERC-404"] do
     {:ok, "--"}
   end
 
   # TODO: remove this clause along with token transfer denormalization
-  defp do_token_transfer_amount(%Token{type: "ERC-20", decimals: nil}, nil, amount, _amounts, _token_ids) do
+  defp do_token_transfer_amount(%Token{type: type, decimals: nil}, nil, amount, _amounts, _token_ids)
+       when type in ["ERC-20", "ERC-404"] do
     {:ok, CurrencyHelper.format_according_to_decimals(amount, Decimal.new(0))}
   end
```
Comment (Qwerty5Uiop) [inline apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:47]: There is one more missing update in clause on line 48

Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/views/tokens/helper.ex:47]: Thanks, processed in https://github.com/blockscout/blockscout/pull/9407/commits/03ab04106b22cb867542ea44d8354725dffccb6e

**Review by Qwerty5Uiop — APPROVED** [review]

**Pull request merged**